### PR TITLE
Add node metrics screen

### LINKED ISSUE #1331: Add gamification to zcashd
In our meeting today, @zookozcash suggested displaying a running tally of various mining and validation metrics on stdout while zcashd is running. The idea is that by giving the user visible numbers that increase the longer they run their node, and a sense of the public good they are doing by running it, they will have an incentive to leave their node running.

Suggested display output:
- ASCII art :smile:
- Thank the user for contributing to the social good
- Let them know whether their miner is running or not
  - Link to FAQ on how to enable
  - (This should reduce the number of user queries we get about how to check this)
- Metrics (all counted since start)
  - Number of days the node has run
  - Number of transactions validated
  - Estimated hash power contribution (if `gen=1`)
  - Number of Equihash solver runs completed (if `gen=1`)
  - Number of blocks won (if `gen=1` and only if at least one block)

This would be tied to a config flag `gamify` which would be enabled by default (so users could disable it by setting `gamify=0` in their `zcash.conf`). It would be completely disabled if `printtoconsole` is set, as the interaction between a scrolling debug log output and an ncurses-style updating output would be... "unpleasant" without further changes to the output logging.

For 1.0 this would be implemented in a Linux-only fashion using direct terminal commands to `std::cout`, to avoid having to link in another dependency. Later, we could switch to using `ncurses` or somthing similar with cross-platform CLI support.

### PR DESCRIPTION
**Pull request by str4d**
Continuation of #1336
Closes #1331

### CODE REVIEW

**Comment by str4d** [comment]
I have rebased this PR on beta 2.

**Comment by daira** [comment]
I rebased this on current master (post-#1578).

**Comment by daira** [comment]
utACK.

**Comment by str4d** [comment]
utACK, looks like @daira added the extra count calls for tromp's solver.

Note that the print statements in tromp's solver will interfere with the metrics screen output. I'll test it now and see if I'd need to fix it (which i can do now).

**Comment by daira** [comment]
It's fine to just comment out those print statements, I think.

**Comment by str4d** [comment]
Yeah, that's what I'll do for now.

**Comment by str4d** [comment]
Okay, it interferes with one of the auto-update lines, but doesn't obscure any information and doesn't break the UI. Still, I'll fix it now :)

**Comment by str4d** [comment]
Found a bug: the solver increment call for tromp's solver was added inside the solution for loop, so it is counting every found solution instead of every run. Rebasing on master to fix.

**Comment by daira** [comment]
Oops, should have checked that more carefully.

**Comment by str4d** [comment]
Rebased on master. Should be good to go :)

**Comment by str4d** [comment]
Rebased on master. Should be good to go :)

File: src/pow/tromp/equi_miner.h [hunk 1]
```diff
@@ -603,7 +603,7 @@ typedef struct {
 void barrier(pthread_barrier_t *barry) {
   const int rc = pthread_barrier_wait(barry);
   if (rc != 0 && rc != PTHREAD_BARRIER_SERIAL_THREAD) {
-    printf("Could not wait on barrier\n");
+//    printf("Could not wait on barrier\n");
```
Comment (daira) [inline src/pow/tromp/equi_miner.h]: This should be a LogPrint instead.

**Comment by daira** [comment]
re-utACK.

File: src/pow/tromp/equi_miner.h (hunk 1 above)
Comment (str4d) [inline src/pow/tromp/equi_miner.h]: We don't actually use that function yet, so we can fix it later.

**Comment by str4d** [comment]
Assigning to @nathan-at-least for a decision, because @zookozcash really really *really* wants this merged, particularly as we also merged #1578 to include a faster miner (although that is not enabled by default, but everyone is going to use it anyway).

**Comment by daira** [comment]
@nathan-at-least said:
> I won't have time today, most likely. I'm comfortable with you deciding. If it still needs my input on Monday I'll look at it early.

My decision is yes, provided it merges and tests cleanly now. @zkbot r+

**Comment by str4d** [comment]
@daira:
> My decision is yes, provided it merges and tests cleanly now.

I just pushed the one-line fix to get the RPC tests running. Should I re-run?

**Comment by str4d** [comment]
Gonna re-run

@zkbot r+

**Comment by daira** [comment]
@zkbot cancel

**Comment by daira** [comment]
Please don't r+ until I've checked something locally.

**Comment by daira** [comment]
OK, ``showmetrics`` is disabled when ``printtoconsole`` is enabled. Just doing a smoke test...

**Comment by daira** [comment]
Smoke-tested ACK. @zkbot r+

**Pull request merged**